tools/c7n_mailer/c7n_mailer/ldap_lookup.py

```diff
@@ -0,0 +1,146 @@
+
+    def get_connection(self, ldap_uri, ldap_bind_user, ldap_bind_password):
+        # note, if ldap_bind_user and ldap_bind_password are None
+        # an anonymous bind will be attempted.
+        try:
+            conn = Connection(
+                ldap_uri, user=ldap_bind_user, password=ldap_bind_password,
+                auto_bind=True,
+                receive_timeout=30,
+                auto_referrals=False,
+            )
+        except LDAPSocketOpenError:
+            conn = None
+        return conn
+
+    def get_email_from_metadata(self, ldap_user_metadata):
+        return ldap_user_metadata.get(self.email_key, None)
+
+    def user_metadata_has_manager(self, ldap_user_metadata):
+        return ldap_user_metadata.get('manager', None)
+
+    def get_metadata_from_dn(self, user_dn):
+        if self.cache_engine:
+            cache_result = self.caching.get(user_dn, None)
+            if cache_result:
+                return cache_result
+        ldap_filter = '(%s=*)' % self.uid_key
+        self.connection.search(user_dn, ldap_filter, attributes=self.attributes)
+        ldap_user_metadata = self.get_dict_from_ldap_object(self.connection.entries[0])
+        if self.cache_engine:
+            self.caching.set(user_dn, ldap_user_metadata)
+        return ldap_user_metadata
+
+    def get_toaddr_from_ldap_metadata(self, ldap_user_to_email_map, ldap_username, email_manager):
+        email_to_addr = []
+        ldap_user_metadata = ldap_user_to_email_map.get(ldap_username)
+        user_email = ldap_user_metadata.get(self.email_key, None)
+        if user_email:
+            email_to_addr.append(user_email)
+        if email_manager:
+            user_manager_dn = ldap_user_metadata.get('manager', None)
+            if user_manager_dn:
+                ldap_user_manager_metadata = ldap_user_to_email_map.get(user_manager_dn, {})
+                user_manager_email = ldap_user_manager_metadata.get(self.email_key, None)
+                if user_manager_email:
+                    email_to_addr.append(user_manager_email)
+        return email_to_addr
+
+    def get_dict_from_ldap_object(self, ldap_user_object):
+        ldap_user_metadata = {attr.key: attr.value for attr in ldap_user_object}
+        ldap_user_metadata['dn'] = ldap_user_object.entry_dn
+        ldap_user_metadata[self.email_key] = ldap_user_metadata[self.email_key].lower()
+        return ldap_user_metadata
+
+    def get_metadata_from_uid(self, uid):
+        if self.uid_regex:
+            # for example if you set ldap_uid_regex in your mailer.yml to "^[0-9]{6}$" then it
+            # would only query LDAP if your string length is 6 characters long and only digits.
+            # re.search("^[0-9]{6}$", "123456")
+            # Out[41]: <_sre.SRE_Match at 0x1109ab440>
+            # re.search("^[0-9]{6}$", "1234567") returns None, or "12345a' also returns None
+            if not re.search(self.uid_regex, uid):
```

Comment:
note, this regex makes it easy to filter away potential queries to AD. So you can require (as an example) that your UID is 6 characters long always, and includes numbers only. ( "^[0-9]{6}$" )

Or you can require an underscore to be in the UID, each unique schema and how they setup their UID can be accommodated to reduce the load on AD.